### PR TITLE
[11.0][ADD] website_lazy_load_image: lazy loading of images

### PR DESCRIPTION
**Pull request by tarteo**
This module loads images on the website lazy which means images are loaded only when the visitor scrolls to it (visible in viewport). This reduces network traffic for your website and your visitors. It uses http://jquery.eisbehr.de/lazy/

cc: @simahawk

### CODE REVIEW

File: website_lazy_load_image/static/src/js/editor.js
Comment (simahawk) [inline website_lazy_load_image/static/src/js/editor.js]: I think we should take care also of background images

File: website_lazy_load_image/readme/CONFIGURE.rst
Comment (simahawk) [inline website_lazy_load_image/readme/CONFIGURE.rst]: The main issue IMO is that you are hacking and writing img on the fly in the editor. This leads to 2 problems:

1. the one you  mentioned
2. all the other images in the website out of the ones you edit are not taken in consideration

The [idea I had in mind](https://github.com/OCA/website/issues/425#issuecomment-365547007) was to transform the final HTML of the whole page so that we can handle this for ALL pages and ALL images in the site.

If you don't like this idea - perfectly fine :wink: - maybe you can make the editor hack conditional based on a config param or flag on the website?

This would leave the door open for an extension of top of it to deal w/ all images.

What do you think?

File: website_lazy_load_image/static/src/js/frontend.js
Comment (simahawk) [inline website_lazy_load_image/static/src/js/frontend.js:65]: are these events you trigger specific to the plugin? If yes I'd prefix them w/ 'lazyload_'

**Review by simahawk — CHANGES_REQUESTED** [review]

File: website_lazy_load_image/readme/CONFIGURE.rst
Comment (tarteo) [reply website_lazy_load_image/readme/CONFIGURE.rst]: Yes, transforming the HTML on transaction is better. But for now I can't spend time on it, this was for a customer and this was fine for them. When the transforming is made I think the client side transforming can be removed.

File: website_lazy_load_image/static/src/js/frontend.js
Comment (tarteo) [reply website_lazy_load_image/static/src/js/frontend.js:65]: Yes it's triggered by the lazyload plugin.

File: website_lazy_load_image/static/src/js/editor.js
Comment (tarteo) [reply website_lazy_load_image/static/src/js/editor.js]: :+1:  will do when I have time.

File: website_lazy_load_image/models/ir_ui_view.py
Comment (simahawk) [inline website_lazy_load_image/models/ir_ui_view.py]: please rename to LAZYLOAD_DEFAULT_SRC and say what kind of image is and how is used in a comment ;)

File: website_lazy_load_image/models/ir_ui_view.py
Comment (simahawk) [inline website_lazy_load_image/models/ir_ui_view.py:15]: docstring documenting why and what we do

File: website_lazy_load_image/models/ir_ui_view.py
Comment (simahawk) [inline website_lazy_load_image/models/ir_ui_view.py]: I'd exclude images having already a class "lazy" and I'd handle background-image styles if any.
The latter can be postponed to a ROADMAP item if you don't have time

File: website_lazy_load_image/models/ir_ui_view.py
Comment (simahawk) [inline website_lazy_load_image/models/ir_ui_view.py:34]: I'd add the class "lazy"

File: website_lazy_load_image/static/src/js/frontend.js
Comment (simahawk) [inline website_lazy_load_image/static/src/js/frontend.js]: filter on .lazy

File: website_lazy_load_image/static/src/js/frontend.js
Comment (simahawk) [inline website_lazy_load_image/static/src/js/frontend.js:95]: as we said at OCA CS: document how you can hook to JS in README and here just place some tech comments (basically what you've explained me yesterday ;))

**Review by simahawk — CHANGES_REQUESTED** [review]
@tarteo tnx a lot! LG. The rest... you know ;)

File: website_lazy_load_image/static/src/js/frontend.js
Comment (tarteo) [reply website_lazy_load_image/static/src/js/frontend.js:95]: I've added this to the js file. I think this is only technical thing so, I don't have a mention in the README.

File: website_lazy_load_image/models/ir_ui_view.py
Comment (tarteo) [reply website_lazy_load_image/models/ir_ui_view.py:34]: I've added lazyload-disable class to let the user exclude images.

File: website_lazy_load_image/models/ir_ui_view.py
Comment (tarteo) [reply website_lazy_load_image/models/ir_ui_view.py:15]: Added it :)

File: website_lazy_load_image/models/ir_ui_view.py
Comment (simahawk) [inline website_lazy_load_image/models/ir_ui_view.py:27]: @tarteo have you tested this? :)
Which reminds me: would be better to have a unit test. (Expected to fail :stuck_out_tongue_winking_eye:)

Comment (tarteo) [reply website_lazy_load_image/models/ir_ui_view.py:27]: Yes, it works :stuck_out_tongue_closed_eyes: I'm currently working on the tests.

Comment (tarteo) [reply website_lazy_load_image/models/ir_ui_view.py:27]: The tests passed :)

File: website_lazy_load_image/tests/test_lazy_load_image.py
Comment (simahawk) [inline website_lazy_load_image/tests/test_lazy_load_image.py]: pls, use `SavepointCase` and replace `setUp` w/ `setUpClass`: will make test faster

File: website_lazy_load_image/models/ir_ui_view.py
Comment (simahawk) [reply website_lazy_load_image/models/ir_ui_view.py:15]: weird... AFAIR `hasclass` is peculiar to odoo's inheritance rendering... :thinking: 
well, TIL :wink: 
@tarteo thanks! Small thing to change in the test but looks good.

File: website_lazy_load_image/tests/test_lazy_load_image.py
Comment (simahawk) [inline website_lazy_load_image/tests/test_lazy_load_image.py]: :thinking:  lemme guess... bad copy/paste session? :rofl:

**Review by simahawk — APPROVED** [review]
thanks for your work!
To me, we miss only squashing on merge...

File: website_lazy_load_image/tests/test_lazy_load_image.py
Comment (tarteo) [reply website_lazy_load_image/tests/test_lazy_load_image.py]: Changed it. More like lazy use of code snippets. :sweat_smile:

**Comment by tarteo** [comment]
Thanks for your intense reviewing :+1:

**Comment by tarteo** [comment]
@OCA/website-maintainers Can somebody review this, please? :)

**Review by astirpe — APPROVED** [review]
Works fine! Code review ok.

**Comment by tarteo** [comment]
Can this be merged or does it needs another review?

**Comment by simahawk** [comment]
@tarteo let's do it. If we have bugs.... we can fix them :wink:

**Pull request merged**
